finality-aleph/src/justification/compatibility.rs

```diff
@@ -57,12 +57,17 @@ enum VersionedAlephJustification {
 }
 
 fn encode_with_version(version: Version, mut payload: Vec<u8>) -> Vec<u8> {
-    let mut result = version.encode();
+    let mut result =
+        Vec::with_capacity(size_of::<Version>() + size_of::<ByteCount>() + payload.len());
+    version.encode_to(&mut result);
     // This will produce rubbish if we ever try encodings that have more than u16::MAX bytes. We
     // expect this won't happen, since we will switch to proper multisignatures before proofs get
     // that big.
-    let num_bytes = payload.len() as ByteCount;
-    result.append(&mut num_bytes.encode());
+    payload
+        .len()
+        .try_into()
+        .unwrap_or(u16::MAX)
+        .encode_to(&mut result);
     result.append(&mut payload);
```

Comment:
In one case you're cloning the `payload` and then, `append` does it again. If you had passed it in via reference (`payload: &Vec<u8>`) then here you could use `result.extend_from_slice(&payload)` and get away without the `clone()`.

Reply:
From what I understand `extend_from_slice` iterates over whole slice, clones each element and pushes it to vec while append moves all data at once. I'm not sure if using it is any better than cloning whole vec and after that moving everything with append.

Reply:
That's a good point -- although you're looking at needlessly copying up to 16kB of data each time. We would have to do some benchmarks but I think allocations that big are more expensive than iteration over a slice where elements live next to each other already.

Reply:
On the other hand we usually have the `V1` here, not `Other`, so optimizing for this case does not change much. Better optimization would be to use `encode_to` instead of `encode`.